{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Comparing assemblies with QUAST\n",
    "\n",
    "The assembly algorithms use different heuristic approaches to reconstruct complete chromosomes, correct errors in the reads and try to solve the problem of the repetitive regions of the genome. This is a challenge when using methods that evaluate the quality of an assembly and compare the different assemblies.\n",
    "\n",
    "[QUAST](https://github.com/ablab/quast) (QUality ASsessment Tool) evaluates and compares genome assemblies and provides various important metrics. It builds comparisons between multiple assemblies of a given genome and can also add the reference genome to the comparison (when available) to extract additional metrics. The QUAST output consists on metrics and graphs that facilitate the interpretation of the assembly results. QUAST also provides metrics to describe structural errors in contigs, reports a list of fully/partially covered genes for each assembly, and presents a series of statistics in a graphic form. However, QUAST does not report scaffolding statistics. The user will need other software to assess this. QUAST also has a [web interface](http://quast.bioinf.spbau.ru/) for comparison of assemblies, with the option to compare against a reference. \n",
    "\n",
    "QUAST adds methods and quality metrics to evaluate the assembly from existing software (Plantagora, GAGE, GeneMark.hmm and GlimmerHMM) generating statistics such as the well-known N50 (minimum contig length to cover at least 50% of the genome) and introducing new statistics like the NA50 (is N50 where the length of aligned blocks is counted instead of the contig lengths). To align the assemblies with a reference genome and evaluate the metrics according to the alignments, QUAST uses the Nucmer aligner. \n",
    "\n",
    "The common output of an assembly tool, such as Canu or Flye, consists on a directory with intermediate files and a final assembly or contigs files in FASTA format. QUAST command can take several contigs files in FASTA for each assembly that the user wants to include in the comparison. The user can also include the genome reference and specify the name of the directory for the output. Using test assemblies computed in previous notebooks, the following provides an example of using QUAST from NanoDJ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "!quast.py -h"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The next command runs the QUAST analysis using the assemblies generated by Canu and Flye in previous notebooks. The <font color='blue'>-R</font> option specifies the reference genome and the <font color='blue'>-o</font> option is to indicate the output directory name:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "!quast.py data/sample/canu_output/sample.contigs.fasta data/sample/flye_output/contigs.fasta \\\n",
    "         -l Canu,Flye \\\n",
    "         -R data/sample/reference/NZ_CP010867.1_Ref.fasta \\\n",
    "         -o sample_quast"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Once QUAST finishes the execution, the output files are stored under the output directory. Integrated in the notebook, QUAST will also show a summary of the genome statistics for the compared assemblies. However, QUAST will also create an HTML report that contains more details that can be viewed using an internet browser. Data tables and plots are also available in individual files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All statistics are based on contigs of size >= 500 bp, unless otherwise noted (e.g., \"# contigs (>= 0 bp)\" and \"Total length (>= 0 bp)\" include all contigs).\n",
      "\n",
      "Assembly                     Canu         Flye       \n",
      "# contigs (>= 0 bp)          48           21         \n",
      "# contigs (>= 1000 bp)       39           21         \n",
      "# contigs (>= 5000 bp)       24           21         \n",
      "# contigs (>= 10000 bp)      14           21         \n",
      "# contigs (>= 25000 bp)      5            16         \n",
      "# contigs (>= 50000 bp)      1            5          \n",
      "Total length (>= 0 bp)       428398       796700     \n",
      "Total length (>= 1000 bp)    423607       796700     \n",
      "Total length (>= 5000 bp)    387929       796700     \n",
      "Total length (>= 10000 bp)   321740       796700     \n",
      "Total length (>= 25000 bp)   190487       690632     \n",
      "Total length (>= 50000 bp)   63190        301257     \n",
      "# contigs                    45           21         \n",
      "Largest contig               63190        65748      \n",
      "Total length                 427823       796700     \n",
      "Reference length             2183395      2183395    \n",
      "GC (%)                       37.20        40.10      \n",
      "Reference GC (%)             35.56        35.56      \n",
      "N50                          19203        41740      \n",
      "N75                          10883        30114      \n",
      "L50                          7            8          \n",
      "L75                          14           13         \n",
      "# misassemblies              0            0          \n",
      "# misassembled contigs       0            0          \n",
      "Misassembled contigs length  0            0          \n",
      "# local misassemblies        19           0          \n",
      "# scaffold gap size mis.     0            0          \n",
      "# unaligned mis. contigs     14           15         \n",
      "# unaligned contigs          2 + 11 part  6 + 15 part\n",
      "Unaligned length             121732       682427     \n",
      "Genome fraction (%)          13.472       5.111      \n",
      "Duplication ratio            1.041        1.024      \n",
      "# N's per 100 kbp            0.00         0.00       \n",
      "# mismatches per 100 kbp     953.63       746.50     \n",
      "# indels per 100 kbp         1963.02      0.00       \n",
      "Largest alignment            38394        354        \n",
      "Total aligned length         302920       114273     \n",
      "NA50                         7758         -          \n",
      "NGA50                        -            -          \n",
      "LA50                         12           -          \n"
     ]
    }
   ],
   "source": [
    "!cat sample_quast/report.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Generated plots are available in the subdirectories under the main output directory. Some of these plots are:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "agalactiae.contigs_GC_content_plot.pdf\tcumulative_plot.pdf\t  gc.icarus.txt\n",
      "Canu_GC_content_plot.pdf\t\tFlye_GC_content_plot.pdf  NGx_plot.pdf\n",
      "contigs_GC_content_plot.pdf\t\tGC_content_plot.pdf\t  Nx_plot.pdf\n",
      "cumulative_plot.pdf  NAx_plot.pdf  NGAx_plot.pdf\n"
     ]
    }
   ],
   "source": [
    "!ls agalactiae_quast/basic_stats/ && ls agalactiae_quast/aligned_stats/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following is a Python code can be used to explore the generated plots listed above directly from the notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <iframe\n",
       "            width=\"700\"\n",
       "            height=\"500\"\n",
       "            src=\"agalactiae_quast/basic_stats/cumulative_plot.pdf\"\n",
       "            frameborder=\"0\"\n",
       "            allowfullscreen\n",
       "        ></iframe>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.lib.display.IFrame at 0x7f0450a78b70>"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import IFrame\n",
    "#Replace first argument of the IFrame function with the path of any PDF plot\n",
    "IFrame(\"agalactiae_quast/basic_stats/cumulative_plot.pdf\", width=700, height=500)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reference\n",
    "\n",
    "[Alexey Gurevich, Vladislav Saveliev, Nikolay Vyahhi, Glenn Tesler; QUAST: quality assessment tool for genome assemblies, Bioinformatics, Volume 29, Issue 8, 15 April 2013, Pages 1072–1075](https://doi.org/10.1093/bioinformatics/btt086)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
