{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Simmulating MinION reads with Nanosim\n",
    "\n",
    "[NanoSim](https://github.com/bcgsc/NanoSim) is a fast and scalable read simulator that captures the technology-specific features of ONT data, and allows for adjustments upon improvement of nanopore sequencing technology.\n",
    "\n",
    "NanoSim proceeds in two steps. In the first step, NanoSim provides a comprehensive alignment-based analysis and generates a set of read profiles (read characterization step). Whereas in the second step, it models the read features such as the base-calling errors of ONT reads and the length distribution (built in the previous step) to produce in silico reads for a given reference genome (simulation step). Although ONT chemistry is still evolving, the statistical models that NanoSim uses are theoretically valid for any version of it. At the moment, it has been tested with both the 1D reads using the R9 chemistry, and the 2D reads using the chemistry R7, R7.3 and R9. NanoSim can simulate either circular or linear genomes and is capable of simulating reads from genomes with multiple chromosomes. An important note is that the runtime and memory requirements of NanoSim are highly dependent on the number of simulated reads and the length of the reference sequence.\n",
    "\n",
    "NanoDJ uses NanoSim-h [Nanosim-h](https://github.com/karel-brinda/NanoSim-H), a fork from the NanoSim project, which permits moving directly to the simulation step using precomputed error profiles instead of investing time on the read characterization step. This alternative may be attractive for those users without access to ONT read-level information data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: nanosim-h [-h] [-v] [-p str] [-o str] [-n int] [-u float] [-m float]\n",
      "                 [-i float] [-d float] [-s int] [--circular] [--perfect]\n",
      "                 [--merge-contigs] [--rnf] [--rnf-add-cigar] [--max-len int]\n",
      "                 [--min-len int] [--kmer-bias int]\n",
      "                 <reference.fa>\n",
      "\n",
      "Program:  NanoSim-H - a simulator of Oxford Nanopore reads.\n",
      "Version:  1.1.0.4\n",
      "Authors:  Chen Yang <cheny@bcgsc.ca> - author of the original software package (NanoSim)\n",
      "          Karel Brinda <kbrinda@hsph.harvard.edu> - author of the NanoSim-H fork\n",
      "\n",
      "positional arguments:\n",
      "  <reference.fa>        reference genome (- for standard input)\n",
      "\n",
      "optional arguments:\n",
      "  -h, --help            show this help message and exit\n",
      "  -v, --version         show program's version number and exit\n",
      "  -p str, --profile str\n",
      "                        error profile - one of precomputed profiles\n",
      "                        ('ecoli_R9_2D', 'ecoli_R9_1D', 'ecoli_UCSC1b',\n",
      "                        'ecoli_R7.3', 'yeast', 'ecoli_R7') or own directory\n",
      "                        with an error profile [ecoli_R9_2D]\n",
      "  -o str, --out-pref str\n",
      "                        prefix of output file [simulated]\n",
      "  -n int, --number int  number of generated reads [10000]\n",
      "  -u float, --unalign-rate float\n",
      "                        rate of unaligned reads [detect from the error\n",
      "                        profile]\n",
      "  -m float, --mis-rate float\n",
      "                        mismatch rate (weight tuning) [1.0]\n",
      "  -i float, --ins-rate float\n",
      "                        insertion rate (weight tuning) [1.0]\n",
      "  -d float, --del-rate float\n",
      "                        deletion rate (weight tuning) [1.0]\n",
      "  -s int, --seed int    initial seed for the pseudorandom number generator (0\n",
      "                        for random) [42]\n",
      "  --circular            circular simulation (linear otherwise)\n",
      "  --perfect             output perfect reads, no mutations\n",
      "  --merge-contigs       merge contigs from the reference\n",
      "  --rnf                 use RNF format for read names\n",
      "  --rnf-add-cigar       add cigar to RNF names (not fully debugged, yet)\n",
      "  --max-len int         maximum read length [inf]\n",
      "  --min-len int         minimum read length [50]\n",
      "  --kmer-bias int       prohibits homopolymers with length >= n bases in\n",
      "                        output reads [6]\n",
      "\n",
      "Examples: nanosim-h --circular ecoli_ref.fasta\n",
      "          nanosim-h --circular --perfect ecoli_ref.fasta\n",
      "          nanosim-h -p yeast --kmer-bias 0 yeast_ref.fasta\n",
      "\n",
      "Notice: the use of `max-len` and `min-len` will affect the read length distributions. If\n",
      "the range between `max-len` and `min-len` is too small, the program will run slowlier accordingly.\n"
     ]
    }
   ],
   "source": [
    "!nanosim-h -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/jovyan/notebooks/data/sample\n",
      "Using profile ecoli_R9_2D from default profiles\n",
      "Simulating unaligned reads\n",
      "100% (12 of 12) |########################| Elapsed Time: 0:00:00 Time:  0:00:00\n",
      "Simulating aligned reads\n",
      "100% (188 of 188) |######################| Elapsed Time: 0:00:04 Time:  0:00:04\n"
     ]
    }
   ],
   "source": [
    "%cd data/sample/\n",
    "!nanosim-h -n 200 reference/NZ_CP010867.1_Ref.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using the QualityControl.ipynb the user can generate distinct plots and tables (read length, GC content, etc) out of the simulated reads:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%load_ext rpy2.ipython\n",
    "\n",
    "from Bio.SeqUtils import GC\n",
    "from Bio import SeqIO\n",
    "\n",
    "#Numpy and pandas to manage our data\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.mlab as mlab\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib\n",
    "\n",
    "#We will use reagular expressions for extract the read id\n",
    "import re\n",
    "\n",
    "#Function to build a pandas dataframe from a FASTA or FASTQ file\n",
    "def read_sequence_file(filename, format):\n",
    "    rows = []\n",
    "    columns = ['id', 'sequence', 'read_length', 'gc_content']\n",
    "    if(format == 'fastq'):\n",
    "        columns.append('avg_quality')\n",
    "    #Building each row with the read and its features\n",
    "    for seq_record in SeqIO.parse(filename, format):\n",
    "        if(format == 'fastq'):\n",
    "            quality_score = np.mean(seq_record.letter_annotations[\"phred_quality\"])\n",
    "        read_id = seq_record.description\n",
    "        if(format == 'fastq'):\n",
    "            rows.append([read_id, seq_record.seq, len(seq_record.seq), GC(seq_record.seq), quality_score])\n",
    "        else:\n",
    "            rows.append([read_id, seq_record.seq, len(seq_record.seq), GC(seq_record.seq)])  \n",
    "        \n",
    "    df = pd.DataFrame(rows, columns=columns)\n",
    "    return df\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAfUAAAFrCAYAAADIJBAiAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvNQv5yAAAG1JJREFUeJzt3XuUZWV95vHvAy0q0tjgtQvUVkQMOoKIitEYA16QEW+DjhoFk3Zca0aMTkwrjJPQSUyW2iYxWV7Q2AgKAQV1RFdU0HhZZgFqt4ggKKgo0NyMtCBegd/8sXeFY1uX0911zql+6/tZ66ze9e6z9/s7b52up/Z7du2dqkKSJO34dpp0AZIkaWEY6pIkNcJQlySpEYa6JEmNMNQlSWqEoS5JUiMMdWmRSLI2yamzrHtKkqvHXVPf96x1zfL8SvLQWdb9YZJzFq46SYMMdWkOSa5M8vMkP01yXZKTk+w26bpGZdS/PFTVaVX19CHqODnJm0ZVh9QqQ12a35FVtRtwIPBo4PgJ16PtkGTZpGuQRsVQl4ZUVdcBn6ELdwCS3DXJ25L8MMn1SU5Mcvd+3R5JPpnkxiQ39ct7D2z74CRfTHJLknOBew9bS5KpJB/p9/39JH8ysG5tkg8n+UC/70uSHDyw/qAkX+/XnZnkQ0nelOQewKeAqX5m4qdJpvrNdpltf7N4apLL+9f9ziTp+355ki/3y0nyD0luSPKTJBcleWSSVwJ/CLy+r+ET/fN/J8kXkmzua3j2wGu6V5JPJLk5yVf71/PlgfWV5FVJLgcu79v+MclV/TYbkvzeFmN4ZpJT+9f8zSQPS3J8X+9VSeadcZDGzVCXhtQH8jOBKwaa3wI8jC7oHwrsBfxFv24n4P3Ag4AHAj8H3jGw7b8AG+jC/K+BY4asYyfgE8A3+v4OA16b5BkDT3s2cAawAjh7ut8kuwAfA04G9gROB54HUFW39q9vU1Xt1j82zbW/OTwLeCxwAPBC4BkzPOfpwJPpxm8F8N+B/6iq9wKnAW/tazgyyV3613wOcF/g1cBpSfbr9/VO4Fbg/nTjONNYPhd4PLB///VX6b5ve9J9L85McreB5x8JfBDYA/g63S90O9GN+V8B75lnDKTxqyofPnzM8gCuBH4K3AIU8DlgRb8udEGyz8DznwB8f5Z9HQjc1C8/ELgNuMfA+n8BTp1l26cAV/fLjwd+uMX644H398trgc8OrNsf+Hm//GTgGiAD678MvGnLfgbWz7q/WWot4EkDX38YOK5ffjnw5X75UOA7wCHATlvs4+Tpmvqvfw+4bvB5dL+QrAV2Bn4N7Dew7k3T/QzUdOg83+ubgAMGXvO5A+uO7N8HO/dfL+/3uWLS71EfPgYfHqlL83tuVS2nC7yHc+c0+X2AXYEN/ZTwZuDTfTtJdk3yniQ/SHIz8CVgRZKdgSm6gL91oJ8fDFnPg+imyDcP9Pt/gPsNPOe6geWfAXfrP0ueAq6pqsE7OV01RJ+z7W/Y5//WyYVV9W90R/zvBK5P8t4ku8+yvyngqqq6Y6DtB3RHzfcBlvGbr2Om1/QbbUlel+TSfup/M3BPfvMjkOsHln8O/Kiqbh/4mplelzRJhro0pKr6It0R5Nv6ph/R/XB/RFWt6B/3rO6kOoDXAfsBj6+q3emOkqE7wr8W2KP/HHvaA4cs5Sq62YAVA4/lVXXEENteC+w1/Rl37wGDL3PIGhZEVf1TVT0GeATdNPyaWerYBDyg/+hh2gPpZh1upJv12Htg3eBr+s/uphf6z8/fQPfRwB5VtQL4Cd33RtphGerS1nk78LQkB/ZHjf8M/EOS+wIk2Wvgs+3ldKG/OcmewAnTO6mqHwBfA/4yyS5JnkQ3xTuMrwA3J3lDkrsn2bk/weyxQ2x7HnA7cGySZUmeAzxuYP31wL2S3HPIWrZZkscmeXz/efmtwC/62qbreMjA0y/on/P6JHdJ8hS68TqjP3r+KLC2nx15OHD0PN0vp/tF4EZgWZK/AGabJZB2GIa6tBWq6kbgA8Cf901voDtx7vx+iv2zdEfn0P0CcHe6I/rz6abmB72E7vPxH9MF/geGrOF2ukA7EPh+v//30U0fz7ftr4DnA6uBzcBLgU8Cv+zXX0b3WfX3+qn9qdn2tQB2p/ul6Ca6qfT/4M5ZkPXA/n0N/6+v+9l0J/L9CHgXcHRfL8CxdK//OrqT206ffk2z+Azdmf7f6fv+BcN9DCEtavnNj9YkLTVJLgBOrKr3T7qWhZLkLcD9q2qovyiQWuGRurTEJPn9JPfvp9+PAR7Fb88i7FCSPDzJo/q/fX8c3UzExyZdlzRuXllJWnr2o/szs92A7wJHVdW1ky1puy2nm3KfAm4A/g74+EQrkibA6XdJkhrh9LskSY3Y4abf733ve9eqVasmXYYkSWOzYcOGH1XVfeZ73g4X6qtWreJrX/vapMuQJGlskgx1xUmn3yVJaoShLklSIwx1SZIaYahLktQIQ12SpEYY6pIkNcJQlySpEYa6JEmNMNQlSWqEoS5JUiMMdUmSGmGoS5LUCENdkqRG7HB3aZNmsnnd6jnXr1izfofoY3ss9vokjZ5H6pIkNcJQlySpEYa6JEmNMNQlSWqEoS5JUiMMdUmSGmGoS5LUCENdkqRGGOqSJDXCUJckqRGGuiRJjTDUJUlqhKEuSVIjDHVJkhphqEuS1AhDXZKkRhjqkiQ1YmShnuSkJDckuXigbV2Sy5JclORjSVaMqn9JkpaaUR6pnwwcvkXbucAjq+pRwHeA40fYvyRJS8rIQr2qvgT8eIu2c6rqtv7L84G9R9W/JElLzSQ/U/9j4FPDPDHJ2iSVpDZt2jTisiRJ2jFNJNSTvBG4DThtmOdX1dqqSlVlampqtMVJkrSDWjbuDpMcAzwLOKyqatz9S5LUqrGGepLDgTcAv19VPxtn35IktW6Uf9J2OnAesF+Sq5OsBt4BLAfOTXJhkhNH1b8kSUvNyI7Uq+rFMzSvH1V/kiQtdV5RTpKkRhjqkiQ1wlCXJKkRhrokSY0w1CVJaoShLklSIwx1SZIaYahLktQIQ12SpEYY6pIkNcJQlySpEYa6JEmNMNQlSWqEoS5JUiMMdUmSGmGoS5LUiGWTLkBLw+Z1q+dcv2LN+jFVMjpL4TVKWtw8UpckqRGGuiRJjTDUJUlqhKEuSVIjDHVJkhphqEuS1AhDXZKkRhjqkiQ1wlCXJKkRhrokSY0w1CVJaoShLklSIwx1SZIaYahLktQIQ12SpEYY6pIkNcJQlySpEYa6JEmNGFmoJzkpyQ1JLh5o2zPJuUku7//dY1T9S5K01IzySP1k4PAt2o4DPldV+wKf67+WJEkLYGShXlVfAn68RfNzgFP65VOA546qf0mSlpplY+7vflV1LUBVXZvkvsNslGQtcALAypUrR1edFq3N61ZPugRJWvR2iBPlqmptVaWqMjU1NelyJElalMYd6tcnWQnQ/3vDmPuXJKlZ4w71s4Fj+uVjgI+PuX9Jkpo1yj9pOx04D9gvydVJVgNvBp6W5HLgaf3XkiRpAYzsRLmqevEsqw4bVZ+SJC1lO8SJcpIkaX6GuiRJjTDUJUlqhKEuSVIjDHVJkhphqEuS1AhDXZKkRhjqkiQ1wlCXJKkRhrokSY0w1CVJaoShLklSIwx1SZIaYahLktQIQ12SpEYY6pIkNWLZpAuQlorN61Zv1/Yr1qxfoEoktcojdUmSGmGoS5LUCENdkqRGGOqSJDXCUJckqRGGuiRJjTDUJUlqhKEuSVIjDHVJkhphqEuS1AhDXZKkRhjqkiQ1wlCXJKkRhrokSY0w1CVJaoShLklSIwx1SZIaYahLktSIiYR6kv+d5JIkFyc5PcndJlGHJEktGXuoJ9kL+BPg4Kp6JLAz8KJx1yFJUmsmNf2+DLh7kmXArsCmCdUhSVIzlo27w6q6JsnbgB8CPwfOqapz5tomyVrgBICVK1eOvEaN3+Z1qye+/xVr1o+0hu016jGStOObxPT7HsBzgAcDU8A9krx0rm2qam1VpaoyNTU1jjIlSdrhTGL6/anA96vqxqr6NfBR4HcnUIckSU2ZRKj/EDgkya5JAhwGXDqBOiRJasrYQ72qLgDOAjYC3+xreO+465AkqTVjP1EOoKpOoD/xTZIkLQyvKCdJUiMMdUmSGmGoS5LUCENdkqRGGOqSJDXCUJckqRFDhXqSQ4dpkyRJkzPskfrbZmhbt5CFSJKk7TPnxWeSPBR4GLB7kiMGVt2T7papkiRpkZjvinJPBF4O3A9YM9B+M/BnI6pJkiRtgzlDvapOAU5J8vKqOnk8JUmSpG0x1LXfq+rkJPsA+wxuU1X/OqrCJEnS1hkq1JP8LfA/6G6RenvfXIChLknSIjHsXdpeCOxTVTePshhJkrTthv2TtmsNdEmSFrdhj9TPS3I6cCbwi+lGP1OXJGnxGDbUH9v/++qBNj9TV1M2r1s96RJGapjXt2LN+jFUImlUhj37/Q9GXYgkSdo+w579fsRM7U6/S5K0eAw7/T54Nbm7AQcCG3H6XZKkRWObpt+T7A/86UgqkiRJ22Sb7qdeVd8CHrXAtUiSpO2wLZ+p70R3Nvw2/UIgSZJGY1s+U78N+C7wgoUvR5IkbSv/pE2SpEYMO/0e4JXAU+kuOnMu8L6qqhHWJkmStsKw0+9vBR4NvL//+hhgX+D1oyhKkiRtvWFD/RnAQVV1G0CSDwMbMNQlSVo0hj2DPXTT7tOqb5MkSYvEsEfqnwE+leRkukB/ed8mSZIWiTlDPcnOwF3pptlfCTyf7gj9bOC9I69OkiQNbb7p9zcDL6mqO6rqxKo6qqr+G7Az8DejL0+SJA1rvlA/gjvPeB/0jn6dJElaJOYL9Tuq6vYtG6vqDuCO0ZQkSZK2xXyhvkuSXbdsTLIb3WftkiRpkZgv1D8EnJJk9+mGJPcE3gecua2dJlmR5KwklyW5NMkTtnVfkiSpM1+o/xXwS+CaJBuTbASuBm4H1m5Hv/8IfLqqHg4cAFy6HfuSJEnM8ydt/RXkXprkoXSXiQ2wsaqu2NYO+6P+J9P9rTtV9SvgV9u6P0mS1Bn2Lm1XANsc5Ft4CHAj8P4kB9BdbvY1VXXrbBskWQucALBy5coFKuNOm9etnnP9ijXrF7zPxWa+MZjPUhgjSVrshr1M7EJaBhwEvLuqHg3cChw31wZVtbaqUlWZmpoaR42SJO1wJhHqVwNXV9UF/ddn0YW8JEnaDmMP9aq6DrgqyX5902HAt8ZdhyRJrRn2hi4L7dXAaUl2Ab4H/NGE6pAkqRkTCfWquhA4eBJ9S5LUqkl8pi5JkkbAUJckqRGGuiRJjTDUJUlqhKEuSVIjDHVJkhphqEuS1AhDXZKkRhjqkiQ1wlCXJKkRhrokSY0w1CVJaoShLklSIwx1SZIaYahLktQIQ12SpEYsm3QBS8HmdavnXL9izfqJ1yBJ2vF5pC5JUiMMdUmSGmGoS5LUCENdkqRGGOqSJDXCUJckqRGGuiRJjTDUJUlqhKEuSVIjDHVJkhphqEuS1AhDXZKkRhjqkiQ1wlCXJKkRhrokSY0w1CVJaoShLklSIwx1SZIaMbFQT7Jzkq8n+eSkapAkqSWTPFJ/DXDpBPuXJKkpEwn1JHsD/xV43yT6lySpRcsm1O/bgdcDy4d5cpK1wAkAK1euHF1Vi9TmdasnXcK8doQatf3m+z6vWLN+TJVImsnYj9STPAu4oao2DLtNVa2tqlRVpqamRlidJEk7rklMvz8ReHaSK4EzgEOTnDqBOiRJasrYQ72qjq+qvatqFfAi4N+q6qXjrkOSpNb4d+qSJDViUifKAVBVXwC+MMkaJElqhUfqkiQ1wlCXJKkRhrokSY0w1CVJaoShLklSIwx1SZIaYahLktQIQ12SpEYY6pIkNcJQlySpEYa6JEmNMNQlSWqEoS5JUiMMdUmSGmGoS5LUCENdkqRGLJt0AYLN61ZPugQJ8L0o7eg8UpckqRGGuiRJjTDUJUlqhKEuSVIjDHVJkhphqEuS1AhDXZKkRhjqkiQ1wlCXJKkRhrokSY0w1CVJaoShLklSIwx1SZIaYahLktQIQ12SpEYY6pIkNcJQlySpEWMP9SQPSPL5JJcmuSTJa8ZdgyRJLVo2gT5vA15XVRuTLAc2JDm3qr41gVokSWrG2I/Uq+raqtrYL98CXArsNe46JElqTapqcp0nq4AvAY+sqpvneN5a4ASAlStXsmnTpgWtY/O61Qu6P0mTsWLN+jnXL8T/9fn6mM98NWzv/jV6w7yPFvr7mGRDVR083/MmdqJckt2AjwCvnSvQAapqbVWlqjI1NTWeAiVJ2sFMJNST3IUu0E+rqo9OogZJklozibPfA6wHLq2qvx93/5IktWoSR+pPBF4GHJrkwv5xxATqkCSpKWP/k7aq+jKQcfcrSVLrvKKcJEmNMNQlSWqEoS5JUiMMdUmSGmGoS5LUCENdkqRGGOqSJDXCUJckqRGGuiRJjTDUJUlqhKEuSVIjDHVJkhphqEuS1AhDXZKkRhjqkiQ1wlCXJKkRyyZdgCS1ZPO61Yu6/xVr1i+JGpYqj9QlSWqEoS5JUiMMdUmSGmGoS5LUCENdkqRGGOqSJDXCUJckqRGGuiRJjTDUJUlqhKEuSVIjDHVJkhphqEuS1AhDXZKkRhjqkiQ1wlCXJKkRhrokSY0w1CVJaoShLklSIyYS6kkOT/LtJFckOW4SNUiS1Jqxh3qSnYF3As8E9gdenGT/cdchSVJrJnGk/jjgiqr6XlX9CjgDeM4E6pAkqSmpqvF2mBwFHF5Vr+i/fhnw+Ko6do5t1gIn9F/+DLh0O8uYAjZt5z40N8d4PBzn0XOMx8NxntuDquo+8z1p2Tgq2UJmaJvzN4uqWgusXbACkqqqqYXan36bYzwejvPoOcbj4TgvjElMv18NPGDg673xtzNJkrbbJEL9q8C+SR6cZBfgRcDZE6hDkqSmjH36vapuS3Is8BlgZ+CkqrpkzGX85Zj7W4oc4/FwnEfPMR4Px3kBjP1EOUmSNBpeUU6SpEYY6pIkNcJQlySpEYa6JEmNMNQlSWqEoS5JUiOWVKh7y9ftl+TKJN9McmGSr/VteyY5N8nl/b979O1J8k/9eF+U5KCB/RzTP//yJMdM6vUsBklOSnJDkosH2hZsTJM8pv+eXdFvO9Olmps3yzivTXJN/36+MMkRA+uO78fs20meMdA+48+R/oJaF/Tj/6H+4lpLSpIHJPl8kkuTXJLkNX277+dxqaol8aC70M13gYcAuwDfAPafdF072gO4Erj3Fm1vBY7rl48D3tIvHwF8iu56/4cAF/TtewLf6//do1/eY9KvbYJj+mTgIODiUYwp8BXgCf02nwKeOenXvIjGeS3wZzM8d//+Z8RdgQf3Pzt2nuvnCPBh4EX98onA/5z0a57AGK8EDuqXlwPf6cfS9/OYHkvpSN1bvo7Oc4BT+uVTgOcOtH+gOucDK5KsBJ4BnFtVP66qm4BzgcPHXfRiUVVfAn68RfOCjGm/bveqOq+6n4gfGNjXkjLLOM/mOcAZVfXLqvo+cAXdz5AZf470R4uHAmf12w9+z5aMqrq2qjb2y7fQ3VFzL3w/j81SCvW9gKsGvr66b9PWKeCcJBuSvLJvu19VXQvdf2rgvn37bGPu92J+CzWme/XLW7brTsf2U78nTU8Ls/XjfC9gc1XdtkX7kpVkFfBo4AJ8P4/NUgr1rb7lq2b0xKo6CHgm8KokT57jubONud+Lbbe1Y+pYz+3dwD7AgcC1wN/17Y7zdkiyG/AR4LVVdfNcT52hzXHeDksp1L3l6wKoqk39vzcAH6Objry+nxaj//eG/umzjbnfi/kt1Jhe3S9v2S6gqq6vqtur6g7gn+nez7D14/wjuqnjZVu0LzlJ7kIX6KdV1Uf7Zt/PY7KUQt1bvm6nJPdIsnx6GXg6cDHdOE6fnXoM8PF++Wzg6P4M10OAn/RTb58Bnp5kj3668+l9m+60IGPar7slySH9575HD+xryZsOmt7z6N7P0I3zi5LcNcmDgX3pTtCa8edI//nu54Gj+u0Hv2dLRv8eWw9cWlV/P7DK9/O4TPpMvXE+6M60/A7d2atvnHQ9O9qD7ozfb/SPS6bHkO7zxM8Bl/f/7tm3B3hnP97fBA4e2Ncf0518dAXwR5N+bRMe19Pppn5/TXcksnohxxQ4mC6svgu8g/7ujEvtMcs4f7Afx4voAmblwPPf2I/Ztxk4w3q2nyP9/4+v9ON/JnDXSb/mCYzxk+imwy8CLuwfR/h+Ht/DW69KktSIpTT9LklS0wx1SZIaYahLktQIQ12SpEYY6pIkNcJQl8Yo3V3uLkvyjf5OVq8YQR9fSPKsGdqfkv7OeqOQZNXApYOn265M8sghtn12knXb0OeMr1Vaqgx1afyOqqoDgBcA70oyNemCFsgq4JXzPWkmVXV2Va1Z2HKkpcdQlyakqi4GbmLghhRJXp/kK0k2JvlEkvv37YclOS/J1/t7Sb9oYJv9+/t4b0xyKnC3YfpPckSSf+9vznNef0Wv6SP6C5O8p7/RyTeS/M7Adn/T38v6giRvGTj6fyewf7/tWQNdvbDf/5VJjp2llpdPbzNf/zN4an/EfkWSvx3Y5xeSvH2mdVKrDHVpQpI8ke6a4d/ov34p8FDgkOpumvOv3HmDkY3Ak6rq0cBTgbcN3FHsg8C7+m3eATx2iL73Af6c7kppjwFeQXc/8GmPAE6sqkf17f+33+5I4FnAAXT3tN53YJtXAd+qqgOr6qiB9l2r6gnAU4A39zf7mM+M/c9if7oxORA4covp+LnWSc1ZNv9TJC2ws/rrVu8DPK+6+3IDPJvuEpgbu9UsA37Sr7sPcFKSfYHbgD2B/ZJ8C3gkXbBTVecn+eYQNTyj7/9LfV8Ay5Lcr1/+dlV9vV8+HziyX/4D4MNVdStAklPofjmYyxl9bVcmuYnuJhyXzbPNbP3P5JTqbnn60yRn0N3X/JNDrJOaY6hL43dUVV2c5AXAqUkeVlXX010H+01VddIM27yb7trkz6+qSvId7pxm35ZrPQf4dFUd/VsruqnuXww03c6dPyuyDf3Ntq+F3gbmrm9bapd2KE6/SxNSVWcC5wDH9U1nA/9relq9v0PYAf26FcCVfaA/jW6anuruVX0x8JJ+m8cB/2WI7s8BDk/yiOmGJPNO29PdiewFSXZNshPwsoF1NwP3HGIfC+1lSZalu3PgC/oah1knNccjdWmyjgc2JHlLVX0wyb2BL/ZT4jsB76L7zP04ujPlj6O7A9ZFA/s4Gnh/kj8FNtBNV8+pqi7vP8Nfn+TuwC7Av9PdWnSu7c5O8rt9Tdf0fU1/tn8R8O0kFwOXbfG5+ihtBD5Ld8LhmVX1ySHXSc3xLm2StkqS5VV1S3+k/j5gU1XNdSLbRCT5AvA2g1xLiUfqkrbWB5KsAu5ONzPw1olWI+k/eaQuSVIjPFFOkqRGGOqSJDXCUJckqRGGuiRJjTDUJUlqxP8H7FG+rPtRPfgAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7fe22ca9a1d0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# ----- Place the path to your reads file here ----- #\n",
    "seq_df = read_sequence_file('simulated.fa', 'fasta')\n",
    "\n",
    "#Setting the plot size\n",
    "matplotlib.rcParams['figure.figsize'] = (9, 8)\n",
    "\n",
    "#We can customize the plot using the pre-defined style called “ggplot” to emulate ggplot (popular R vizualization package) style \n",
    "plt.style.use('seaborn-notebook')\n",
    "\n",
    "#On the plt.hist parameters we can set the number of bins and the bar color using labels as 'blue', 'red', 'yellow'... or using a RGB code.\n",
    "#CHANGE 'read_length' FOR 'gc_content' TO PLOT GC CONTENT INSTAD OF READ LENGTH \n",
    "plt.hist(seq_df['read_length'], bins=50, color='#ef8a62')\n",
    "plt.title('Read length histogram')\n",
    "plt.xlabel('Read length in bp')\n",
    "plt.ylabel('Count')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Nanosim regular version usage\n",
    "\n",
    "In the regular version of NanoSim, the two stages need to be executed separately. For the read characterization step, NanoSim takes a training read set in FASTA format and use minimap2 or LAST as alignment algorithm to characterize base-calling errors. An alignment file in SAM/MAF format can be provided instead of the reads. A set of read profiles is then generated and used as input in the simulation step.\n",
    "\n",
    "Ready to use read profiles are available and can be downloaded from [Supplementary data of NanoSim](ftp://ftp.bcgsc.ca/supplementary/NanoSim/). These profiles allow users to move directly to the simulation step.This is available since the last version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./read_analysis.py <options>\n",
      "<options>: \n",
      "-h : print usage message\n",
      "-i : training ONT real reads, must be fasta files\n",
      "-r : reference genome of the training reads\n",
      "-a : Aligner to be used: minimap2 or LAST, default = 'minimap2\n",
      "-m : User can provide their own alignment file, with maf or sam extension, can be omitted\n",
      "-b : number of bins (for development), default = 20\n",
      "-t : number of threads for LAST alignment, default = 1\n",
      "-o : The prefix of output file, default = 'training'\n",
      "--no_model_fit : Skip the model fitting step\n"
     ]
    }
   ],
   "source": [
    "!read_analysis.py -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2018-11-13 16:46:16: Read pre-process and unaligned reads analysis\n",
      "2018-11-13 16:46:18: Alignment with minimap2\n",
      "[M::mm_idx_gen::0.215*0.43] collected minimizers\n",
      "[M::mm_idx_gen::0.306*0.42] sorted minimizers\n",
      "[M::main::0.306*0.42] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.325*0.42] mid_occ = 8\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.344*0.41] distinct minimizers: 396667 (98.51% are singletons); average occurrences: 1.029; average spacing: 5.347\n",
      "[M::worker_pipeline::6.023*0.85] mapped 3720 sequences\n",
      "[M::main] Version: 2.13-r858-dirty\n",
      "[M::main] CMD: minimap2 --cs -ax map-ont reference/NZ_CP010867.1_Ref.fasta sample_processed.fasta\n",
      "[M::main] Real time: 6.031 sec; CPU: 5.100 sec; Peak RSS: 0.310 GB\n",
      "2018-11-13 16:46:25: Aligned reads analysis\n",
      "2018-11-13 16:46:26: match and error models\n",
      "2018-11-13 16:46:29: Model fitting\n",
      "2018-11-13 16:46:29: Mismatch fitting start\n",
      "Mismatch parameters:  [ 0.13994987  0.77703295  0.19364047] 0.000213784733259\n",
      "2018-11-13 16:49:33: Mismatch fitting done\n",
      "2018-11-13 16:49:33: Insertion fitting start\n",
      "/home/jovyan/NanoSim-2.1.0/src/mixed_model.py:25: RuntimeWarning: overflow encountered in power\n",
      "  wei_cdf = 1 - np.exp(-1 * np.power(x / l, k))\n"
     ]
    }
   ],
   "source": [
    "!read_analysis.py -i reads.fasta -r reference/NZ_CP010867.1_Ref.fasta -o sample"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Simulation stage takes read profiles and reference as input. The output of this stage consists of a FASTA file with the simulated reads."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./simulator.py [command] <options>\n",
      "[command] circular | linear\n",
      "Do not choose 'circular' when there is more than one sequence in the reference\n",
      "<options>: \n",
      "-h : print usage message\n",
      "-r : reference genome in fasta file, specify path and file name, REQUIRED\n",
      "-c : The prefix of training set profiles, same as the output prefix in read_analysis.py, default = training\n",
      "-o : The prefix of output file, default = 'simulated'\n",
      "-n : Number of generated reads, default = 20,000 reads\n",
      "--max_len : Maximum read length, default = Inf\n",
      "--min_len : Minimum read length, default = 50\n",
      "--perfect: Output perfect reads, no mutations, default = False\n",
      "--KmerBias: prohibits homopolymers with length >= n bases in output reads, default = 6\n",
      "--seed: manually seeds the pseudo-random number generator, default = None\n"
     ]
    }
   ],
   "source": [
    "!simulator.py -H"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "!simulator.py circular -r reference/NZ_CP010867.1_Ref.fasta -c sample"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reference:\n",
    "\n",
    "[1] Yang C., Chu J., Warren R.L, Birol I. NanoSim: nanopore sequence read simulator based on statistical characterization. GigaScience, Volume 6, Issue 4, 1 April 2017, gix010. DOI: https://doi.org/10.1093/gigascience/gix010\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
